### ISSUE TITLE
Optimisation of resources for the workflow

### ISSUE THREAD

**Issue by ypriverol** [bigbio/quantms#455]
### Description of feature

Currently, quantms have seven major categories for resources or processes: 

```nextflow
withLabel:process_single {
        cpus   = { 1                   }
        memory = { 6.GB * task.attempt }
        time   = { 4.h  * task.attempt }
    }
    withLabel:process_low {
        cpus   = { 4     * task.attempt }
        memory = { 12.GB * task.attempt }
        time   = { 6.h   * task.attempt }
    }
    withLabel:process_very_low {
        cpus   = { 2     * task.attempt}
        memory = { 4.GB  * task.attempt}
        time   = { 3.h   * task.attempt}
    }
    withLabel:process_medium {
        cpus   = { 8     * task.attempt }
        memory = { 36.GB * task.attempt }
        time   = { 8.h   * task.attempt }
    }
    withLabel:process_high {
        cpus   = { 12    * task.attempt }
        memory = { 72.GB * task.attempt }
        time   = { 16.h  * task.attempt }
    }
    withLabel:process_long {
        time   = { 20.h  * task.attempt }
    }
    withLabel:process_high_memory {
        memory = { 200.GB * task.attempt }
    }
```

However, some of my current analyses showed that resource usage, for example, for DIA analysis, could be optimized much more at the process level. See some results from my analyses. 

### Dataset: PXD030304

**CPU Usage**: 

![Screenshot 2024-12-05 at 09 08 27](https://github.com/user-attachments/assets/f5f1f412-d090-4da6-9dbe-147f01515b30)

**Memory Usage**: 

![Screenshot 2024-12-05 at 09 08 43](https://github.com/user-attachments/assets/dca9cb4a-0f6b-463a-aa02-a5c1dff1193d)

**IO Usage**: 

![Screenshot 2024-12-05 at 09 08 55](https://github.com/user-attachments/assets/bd0e53a6-1d78-4924-b860-a337cad235ed)

**Comment by jpfeuffer** [comment]
My plan was always to use the results of your thousands of runs to learn a simple regression model for each step. But I am not sure if you ever saved the execution logs.

**Comment by ypriverol** [comment]
I did it for most of the runs. However, you don't really need a huge data to be able to learn simple things. Some conclusions, easy ones: 

- `samplecheet_check` and `sdrf_parsing` are way-way over their memory requirements, they can now go easily to 1GB memory, currently we are given to them 6GB.

**Comment by jpfeuffer** [comment]
Well, yes, but I wasnt talking about those easy things. Of course you can add smaller labels for those.

**Comment by ypriverol** [comment]
I think the orher ones depends heavily of the mzML size, number of MS and MS/MS I guess, even type of instrument, or file size.

**Comment by jpfeuffer** [comment]
That's why I said learning from your results..

**Comment by jpfeuffer** [comment]
All this information is available when starting a run..

**Comment by jpfeuffer** [comment]
Would be a unique and potentially publishable feature of the pipeline.
There is still the retry functionality if the resources were not enough. But I assume there should be some very informative features that allow for a very accurate prediction of resource usage.

**Comment by ypriverol** [comment]
Yes, the idea is to optimize the pipeline for each process for 80% of the runs, if the 20 fails, it can go to the next retry. Before doing the research we have to think if is needed to have the info inside the files, MS and MS/MS. because if for the model that information is needed, then we will need to block all process until mzml_statistics finish?

**Comment by jpfeuffer** [comment]
Yes that is true.

**Comment by ypriverol** [comment]
This is DDA-TMT dataset PXD010557: 

**Memory Usage**: 

![Screenshot 2024-12-05 at 09 51 00](https://github.com/user-attachments/assets/49fa8c6f-7ecf-4b02-8939-498510a10e6d)

**CPU Usage**:

![Screenshot 2024-12-05 at 09 51 11](https://github.com/user-attachments/assets/d860e60d-2168-4cb6-971b-d9c9a7f7c4ab)

**IO Usage**: 

![Screenshot 2024-12-05 at 09 51 23](https://github.com/user-attachments/assets/64ec5cb7-47aa-4ef1-802a-68fd36b932f4)

**Comment by jpfeuffer** [comment]
I think we cannot predict CPU usage. We need to know from the implementation if it benefits from multiple cores.
Depending on the implementation multiple cores also means a bit more RAM because more data is loaded at the same time or copies are made for thread-safety

**Comment by timosachsenberg** [comment]
You can also subsample and average statistics from some files to get a much better idea.

**Comment by josenimo** [comment]
@ypriverol @jpfeuffer 
I am restarting this conversation, particularly for .d files on DIANN (DIA) runs.
Currently, `BIGBIO_QUANTMS:QUANTMS:DIA:PRELIMINARY_ANALYSIS` asks for 72gb of RAM. 
I am actually not sure how much is needed, but my raw file is 3.5gb, and my speclib is 2gb.
Now I am testing locally so it breaks my run, but also for the HPC asking that much RAM might be a bottleneck in some systems when running 100s of samples.

thoughts?

PS, you fixes worked!!

**Comment by josenimo** [comment]
I have limited nextflow knowledge, but could one ask dynamically based on file sizes?

**Comment by ypriverol** [comment]
A couple of things here; you can optimize this in multiple ways. First, you can pass the params about how much memory do you want in that step; (honestly, I think in a local machine this will not work). The other thing is that you can pass the # of files that you want to do the preliminary analysis on; doesn't need to be the 100 files; it could be 40 for example; 50.  

You can start by passing 

```
process { 
   withLabel:process_high {
        cpus   = { 12    * task.attempt }
        memory = { 30.GB * task.attempt }
        time   = { 16.h  * task.attempt }
}
```

And pass this as extra parameter to the pipeline

**Comment by jpfeuffer** [comment]
I think the solution for your local mac is to set maxMemory and maxCPUs. nf-core configs (like ours) will automatically cap to that value.

**Comment by josenimo** [comment]
awesome, thank you both, now trying it out in HPC, closer to reality

**Comment by ypriverol** [comment]
Just to be sure, all the Mac analysis, using emulators for all the DIANN, is quite bad. Im trying to figure it out if we can produce a container from DIANN that do not need to use arm emulator which is extremely slow. I always recommend to use quantms with a few files in the machine but quantms is fundamental for HPC and could.

**Comment by josenimo** [comment]
Yeah, I understand, and I think that is the correct way to prioritize. 
I just like to test things locally before adding the extra layer of complexity.
